### PR TITLE
Update platform support and require symengine

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit updates our platform support matrix to reflect upcoming changes. The first is that in Rust 1.74 the Rust programming language is raising their minimum support macOS version to 10.12, so Qiskit is raising it's supported version of macOS to match this. The second change is making symengine a hard requirement. We previously had symengine as a requirement only on platforms that had precompiled packages available. But, the percentage of our user base that runs qiskit on those platforms is very small, and maintaining dual support for symengine and sympy adds a lot of complexity around managing the dependencies. This commit promotes symengine to a hard requirement for all users regardless of platform. As a result Linux i686 and 32 bit Windows for Python < 3.10 has been downgraded to tier 3 support as you'll need a C++ to install Qiskit on that platform now (regardless of Python version).

### Details and comments

This PR is on hold until after the 0.45.0 release. These platform changes don't apply to the 0.45.0 release, I just wanted to push them up earlier so we can debug any potential issues prior well in advance of making the change.

### CODE REVIEW

**Comment by mtreinish** [comment]
This should be ready to go now, I fixed the bug with symengine in: https://github.com/Qiskit/qiskit/pull/10902/commits/4bdcc20f530ea3e2d0061da3701eaaf5a84e618b. I'm going to open a standalone PR for that though so we can backport it to 0.45.1 (I'll remove the release note once I do).

**Comment by mtreinish** [comment]
Rust 1.74 released yesterday so the importance on this PR is increased a bit since whenever we tag a release we'll be pulling that in and won't be able to build binaries that are compatible with macOS 10.9. I do think we should wait for #11261 first though and then rebase this on top of that.

File: releasenotes/notes/platform-support-f7f693aaf5dec044.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+upgrade:
+  - |
+    The ``symengine`` library is now a hard requirement for all platforms.
+    Previously, ``symengine`` was required only on platforms that had
+    pre-compiled packages available and ``sympy`` would be used as a fallback
+    if it wasn't installed. This split requirements was resulting in increased
```
Comment (ElePT) [inline releasenotes/notes/platform-support-f7f693aaf5dec044.yaml]:
```suggestion
    if it wasn't installed. These split requirements were resulting in increased
```

File: releasenotes/notes/platform-support-f7f693aaf5dec044.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    complexity as it was necessary to determine which libraries were installed
```
Comment (ElePT) [inline releasenotes/notes/platform-support-f7f693aaf5dec044.yaml]:
```suggestion
    complexity, as it was necessary to determine which libraries were installed
```

File: releasenotes/notes/platform-support-f7f693aaf5dec044.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    to debug an issue. By requiring symengine for all systems greatly decreases
```
Comment (ElePT) [inline releasenotes/notes/platform-support-f7f693aaf5dec044.yaml]:
```suggestion
    to debug an issue. Requiring ``symengine`` for all systems greatly decreases
```

File: releasenotes/notes/platform-support-f7f693aaf5dec044.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    the complexity and optimizes Qiskit for higher performance. However, for
```
Comment (ElePT) [inline releasenotes/notes/platform-support-f7f693aaf5dec044.yaml]:
```suggestion
    the complexity and optimizes Qiskit for higher performance. However,
```

File: releasenotes/notes/platform-support-f7f693aaf5dec044.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    users on i686 Linux, 32 bit Windows, and s390x Linux (the platforms without
+    precompiled packages on PyPI) will need to build symengine from source.
+  - |
+    Support for 32 bit platforms, i686 Linux and 32 bit Windows, on
+    Python < 3.10 has been downgraded from Tier 2 to Tier 3, as documented in
+    the :ref:`platform_support` page. This was required due to making
+    ``symengine`` required for all users and needing to install symengine from
+    source on these platforms.
```
Comment (ElePT) [inline releasenotes/notes/platform-support-f7f693aaf5dec044.yaml]:
```suggestion
    the :ref:`platform_support` page. This is a consequence of making
    ``symengine`` required for all users, as there is a lack of pre-compiled packages 
    available for these platforms, so users will need to build symengine from
    source.
```

File: releasenotes/notes/platform-support-f7f693aaf5dec044.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  - |
+    For macOS the minimum version of macOS is now 10.12. Previously, the
```
Comment (ElePT) [inline releasenotes/notes/platform-support-f7f693aaf5dec044.yaml]:
```suggestion
    For macOS users, the minimum version of macOS is now 10.12. Previously, the
```

File: requirements.txt [hunk 7]
```diff
@@ -8,7 +8,4 @@ dill>=0.3
 python-dateutil>=2.8.0
 stevedore>=3.0.0
 typing-extensions; python_version<'3.11'
-# symengine pinning needed due lowered precision handling complex
-# multiplication in version 0.10 wich breaks parameter assignment test
-# (can be removed once issue is fix)
-symengine>=0.9, <0.10; platform_machine == 'x86_64' or platform_machine == 'aarch64' or platform_machine == 'ppc64le' or platform_machine == 'amd64' or platform_machine == 'arm64'
+symengine>=0.9, <0.10
```
Comment (ElePT) [inline requirements.txt:11]: should we keep the comment explaining why we are pinning symengine to <0.10?

**Review by ElePT — COMMENTED** [review]
Sorry for the chain of inline comments on the reno, I should have probably read it all at once and made a single suggestion. But these are all minor, overall LGTM.

File: requirements.txt (hunk 7 above)
Comment (mtreinish) [reply requirements.txt:11]: Yeah, normally I'd agree and add it back. But, https://github.com/Qiskit/qiskit/pull/11262 uncaps the symengine version. So I don't think it matters too much anymore.

**Review by ElePT — APPROVED** [review]

**Pull request merged**